docs/user/performance.rst

```diff
@@ -0,0 +1,122 @@
+downsides.
+
+First make sure that there is any significant slowdown
+------------------------------------------------------
+
+Neural nets are notoriously slow to train. Therefore, if your training takes a
+lot of time, that doesn't automatically mean that the skorch overhead is at
+fault. Maybe the training would take the same time without skorch. If you have
+some measurements about training the same model without skorch, first make sure
+that this points to skorch being the culprit before trying to optimize using the
+mitigations described below. If it turns out skorch is not the culprit, look
+into optimizing the performance of PyTorch code in general.
+
+Many people use skorch for hyper-parameter search. Remember that this implies
+fitting the model repeatedly, thus a long run time is expected. E.g. if you run
+a grid search on two hyper-parameters, each with 10 variants, and 5 splits,
+there will actually be 10 x 10 x 5 fit calls, so expect the process to take
+approximately 500 times as long as a single model fit. Increase the verbosity on
+the grid search to get a better idea on the progress (e.g. ``GridSerachCV(...,
+verbose=3)``).
+
+Turning off verbosity
+---------------------
+
+By default, skorch produces a print log of the training progress. This is useful
+for checking the training progress, monitor overfitting, etc. If you don't need
+these diagnostics, you can turn them off via the ``verbose`` parameter. This
+way, printing is deactivated, saving time on i/o. You can still access the
+diagnostics through the ``history`` attribute after training has finished.
+
+.. code:: python
+
+   net = NeuralNet(..., verbose=0)  # turn off verbosity
+   net.fit(X, y)
+   train_loss = net.history[..., 'train_loss']  # access history as usual
+
+Disabling callbacks all together
+--------------------------------
+
+If you don't need any callbacks at all, turning them off can be potential time
+saver. Callbacks present the most significant "extra" that skorch provides over
+pure PyTorch, hence they might add a lot of overhead for small workloads. By
+turning them off, you lose their functionality, though. It's up to you to
+determine if that's a worthwhile trade-off or not. For instance, in contrast to
+just turning down verbosity, you will no longer have access to useful
+diagnostics in the ``history`` attribute.
+
+.. code:: python
+
+   # skorch version 0.10 or later:
+   net = NeuralNet(..., callbacks='disable')
+   net.fit(X, y)
+   print(net.history)  # no longer contains useful diagnostics
+
+   # skorch version 0.9 or earlier
+   net = NeuralNet(...)
+   net.initialize()
+   net.callbacks_ = []  # manually remove all callbacks
+   net.fit(X, y)
+
+Instead of turning off all callbacks, you can also turn off specific callbacks,
+including default callbacks. This way, you can decide which ones to keep and
+which ones to get rid off. Typically, callbacks that calculate some kind of
+metric tend to be slow.
```

Comment:
I am thinking this is an performance issue because data has to be moved to the CPU to compute metrics.

Reply:
Indeed. However, I ran my tests on CPU and there was still a measurable overhead, so it's not solely that.